rs/sns/governance/src/types.rs

```diff
@@ -2527,6 +2537,244 @@ impl From<MintSnsTokens> for Action {
+            } => {
+                validate_chunked_wasm(env, wasm_module_hash, *store_canister_id, chunk_hashes_list)
+                    .await
+            }
+        }
+    }
+
+    pub fn info(&self) -> String {
+        match self {
+            Self::Bytes(bytes) => {
+                let canister_wasm_sha256 = {
+                    let mut state = Sha256::new();
+                    state.write(&bytes[..]);
+                    let sha = state.finish();
+                    sha.to_vec()
+                };
+                format!(
+                    "Embedded module with {} bytes and SHA256 `{}`.",
+                    bytes.len(),
+                    format_full_hash(&canister_wasm_sha256)
+                )
+            }
+            Self::Chunked {
+                wasm_module_hash,
+                store_canister_id,
+                chunk_hashes_list,
+            } => {
+                format!(
+                    "Remote module stored on canister {} with SHA256 `{}`. \
+                     Split into {} chunks:\n  - {}",
+                    store_canister_id.get(),
+                    format_full_hash(wasm_module_hash),
+                    chunk_hashes_list.len(),
+                    chunk_hashes_list
+                        .iter()
+                        .map(|chunk_hash| { format!("`{}`", format_full_hash(chunk_hash)) })
+                        .collect::<Vec<_>>()
+                        .join("\n  - "),
+                )
+            }
+        }
+    }
+}
+
+impl TryFrom<&UpgradeSnsControlledCanister> for Wasm {
+    type Error = String;
+
+    fn try_from(upgrade: &UpgradeSnsControlledCanister) -> Result<Self, Self::Error> {
+        const ERR_PREFIX: &str = "Invalid UpgradeSnsControlledCanister";
+
+        match (
+            &upgrade.new_canister_wasm[..],
+            &upgrade.chunked_canister_wasm,
+        ) {
+            (
+                [],
+                Some(ChunkedCanisterWasm {
+                    wasm_module_hash,
+                    store_canister_id,
+                    chunk_hashes_list,
+                }),
+            ) => {
+                let Some(store_canister_id) = store_canister_id else {
+                    return Err(format!(
+                        "{ERR_PREFIX}.chunked_canister_wasm.store_canister_id must be \
+                             specified."
+                    ));
+                };
+
+                let store_canister_id = CanisterId::try_from_principal_id(*store_canister_id)
+                    .map_err(|err| {
+                        format!("{ERR_PREFIX}.chunked_canister_wasm.store_canister_id: {err}")
+                    })?;
+
+                Ok(Self::Chunked {
+                    wasm_module_hash: wasm_module_hash.clone(),
+                    store_canister_id,
+                    chunk_hashes_list: chunk_hashes_list.clone(),
+                })
+            }
+            (bytes, None) => Ok(Self::Bytes(bytes.to_vec())),
```

Comment:
Note that this could do a substantial amount of copying. To avoid this, it might make sense to switch from reference to consuming (that way, data is moved instead of being copied).

Reply:
Good catch. 

Currently, this copying can't be easily avoided because the whole call stack would need to change, starting from 

```
validate_and_render_action(&proposal.action, ...)
```

However, note that these bytes are limited to approximately 2 MiB, so this shouldn't cause extreme disruption. 

I will try to refactor the code to avoid cloning the proposal action outside of this PR.